bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java

```diff
@@ -0,0 +1,926 @@
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.qbus.internal.QbusBridgeHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonParseException;
+
+/**
+ * The {@link QbusCommunication} class is able to do the following tasks with Qbus
+ * systems:
+ * <ul>
+ * <li>Start and stop TCP socket connection with Qbus Server.
+ * <li>Read all the outputs and their status information from the Qbus Controller.
+ * <li>Execute Qbus commands.
+ * <li>Listen to events from Qbus.
+ * </ul>
+ *
+ * A class instance is instantiated from the {@link QbusBridgeHandler} class initialization.
+ *
+ * @author Koen Schockaert - Initial Contribution
+ */
+
+@NonNullByDefault
+public final class QbusCommunication {
+
+    private final Logger logger = LoggerFactory.getLogger(QbusCommunication.class);
+
+    @Nullable
+    private Socket qSocket;
+    @Nullable
+    private PrintWriter qOut;
+    @Nullable
+    private BufferedReader qIn;
+
+    private boolean listenerStopped;
+    private boolean qEventsRunning;
+
+    private Gson gsonOut = new Gson();
+    private Gson gsonIn;
+    @Nullable
+    private String CTD;
+
+    private Boolean CTDConnected = false;
+
+    private final Map<Integer, QbusBistabiel> bistabiel = new HashMap<>();
+    private final Map<Integer, QbusScene> scene = new HashMap<>();
+    private final Map<Integer, QbusDimmer> dimmer = new HashMap<>();
+    private final Map<Integer, QbusRol> rol = new HashMap<>();
+    private final Map<Integer, QbusThermostat> thermostat = new HashMap<>();
+    private final Map<Integer, QbusCO2> co2 = new HashMap<>();
+
+    @Nullable
+    private QbusBridgeHandler bridgeCallBack;
+
+    /**
+     * Constructor for Qbus communication object, manages communication with
+     * Qbus Server.
+     *
+     */
+    public QbusCommunication() {
+        GsonBuilder gsonBuilder = new GsonBuilder();
+        gsonBuilder.registerTypeAdapter(QbusMessageBase.class, new QbusMessageDeserializer());
+        gsonIn = gsonBuilder.create();
+    }
+
+    /**
+     * Start communication with Qbus Server, run through initialization and start thread listening
+     * to all messages coming from Qbus.
+     *
+     * @param addr : IP-address of Qbus Server
+     * @param port : Communication port of Qbus server
+     *
+     */
+
+    public synchronized void startCommunication() {
+        QbusBridgeHandler handler = bridgeCallBack;
+        CTDConnected = false;
+        try {
+
+            for (int i = 1; qEventsRunning && (i <= 5); i++) {
+                Thread.sleep(1000);
+            }
+            if (qEventsRunning) {
+                logger.error("Starting from thread {}, but previous connection still active after 5000ms",
+                        Thread.currentThread().getId());
```

Comment:
You could add this message to the IOException. Then, you only need one logging statement for all exceptions in the catch block.